libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp

```diff
@@ -44,64 +44,63 @@ const AP_Param::GroupInfo AP_BoardConfig_CAN::var_info[] = {
 
 #if MAX_NUMBER_OF_CAN_INTERFACES > 0
     // @Group: P1_
-    // @Path: ../AP_BoardConfig/canbus.cpp
-    AP_SUBGROUPINFO(_var_info_can[0], "P1_", 1, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_if_var_info),
+    // @Path: ../AP_BoardConfig/canbus_interface.cpp
+    AP_SUBGROUPINFO(_interfaces[0], "P1_", 1, AP_BoardConfig_CAN, AP_BoardConfig_CAN::Interface),
 #endif
 
 #if MAX_NUMBER_OF_CAN_INTERFACES > 1
     // @Group: P2_
-    // @Path: ../AP_BoardConfig/canbus.cpp
-    AP_SUBGROUPINFO(_var_info_can[1], "P2_", 2, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_if_var_info),
+    // @Path: ../AP_BoardConfig/canbus_interface.cpp
+    AP_SUBGROUPINFO(_interfaces[1], "P2_", 2, AP_BoardConfig_CAN, AP_BoardConfig_CAN::Interface),
 #endif
 
 #if MAX_NUMBER_OF_CAN_INTERFACES > 2
     // @Group: P3_
-    // @Path: ../AP_BoardConfig/canbus.cpp
-    AP_SUBGROUPINFO(_var_info_can[2], "P3_", 3, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_if_var_info),
+    // @Path: ../AP_BoardConfig/canbus_interface.cpp
+    AP_SUBGROUPINFO(_interfaces[2], "P3_", 3, AP_BoardConfig_CAN, AP_BoardConfig_CAN::Interface),
 #endif
 
 #if MAX_NUMBER_OF_CAN_DRIVERS > 0
     // @Group: D1_
     // @Path: ../AP_BoardConfig/canbus_driver.cpp
-    AP_SUBGROUPINFO(_var_info_can_protocol[0], "D1_", 4, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_driver_var_info),
+    AP_SUBGROUPINFO(_drivers[0], "D1_", 4, AP_BoardConfig_CAN, AP_BoardConfig_CAN::Driver),
 #endif
 
 #if MAX_NUMBER_OF_CAN_DRIVERS > 1
     // @Group: D2_
     // @Path: ../AP_BoardConfig/canbus_driver.cpp
-    AP_SUBGROUPINFO(_var_info_can_protocol[1], "D2_", 5, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_driver_var_info),
+    AP_SUBGROUPINFO(_drivers[1], "D2_", 5, AP_BoardConfig_CAN, AP_BoardConfig_CAN::Driver),
 #endif
 
 #if MAX_NUMBER_OF_CAN_DRIVERS > 2
     // @Group: D3_
     // @Path: ../AP_BoardConfig/canbus_driver.cpp
-    AP_SUBGROUPINFO(_var_info_can_protocol[2], "D3_", 6, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_driver_var_info),
+    AP_SUBGROUPINFO(_drivers[2], "D3_", 6, AP_BoardConfig_CAN, AP_BoardConfig_CAN::Driver),
 #endif
 
     AP_GROUPEND
 };
 
-int8_t AP_BoardConfig_CAN::_st_driver_number[MAX_NUMBER_OF_CAN_INTERFACES];
-int8_t AP_BoardConfig_CAN::_st_can_debug[MAX_NUMBER_OF_CAN_INTERFACES];
+AP_BoardConfig_CAN *AP_BoardConfig_CAN::_singleton;
 
-void AP_BoardConfig_CAN::init()
+AP_BoardConfig_CAN::AP_BoardConfig_CAN()
 {
-    for (uint8_t i = 0; i < MAX_NUMBER_OF_CAN_INTERFACES; i++)
-    {
-        _st_driver_number[i] = (int8_t) _var_info_can[i]._driver_number;
-        _st_can_debug[i] = (int8_t) _var_info_can[i]._can_debug;
+    AP_Param::setup_object_defaults(this, var_info);
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    if (_singleton != nullptr) {
+        AP_HAL::panic("AP_Proximity must be singleton");
```

Comment:
AP_Proximity?

Reply:
Ups, copy paste done wrong :disappointed: I'll fix it.